src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -47,73 +82,145 @@ public SemanticTokensRangeHandler(IGlobalOptionService globalOptions)
-                {
-                    // We don't want to hang on to the compilation since this can be very expensive,
-                    // but we do want to mark the compilation as being successfully retrieved.
-                    if (compilationTask.Result is not null)
-                    {
-                        _projectIdToCompilation[project.Id] = Task.FromResult<Compilation?>(null);
-                    }
+                // If this checksum is hte same as the last computed result, no need to continue, we would not produce a
+                // different compilation.
+                if (ChecksumIsUnchanged_NoLock(project, projectChecksum))
+                    return;
 
-                    return true;
+                if (!_projectIdToEventSource.TryGetValue(project.Id, out var eventSource))
+                {
+                    eventSource = new CompilationAvailableEventSource(_asyncListener);
+                    _projectIdToEventSource.Add(project.Id, eventSource);
                 }
-            }
 
-            var workspaceStatusService = project.Solution.Workspace.Services.GetRequiredService<IWorkspaceStatusService>();
-            var isFullyLoaded = await workspaceStatusService.IsFullyLoadedAsync(cancellationToken).ConfigureAwait(false);
+                eventSource.EnsureCompilationAvailability(project, () => OnCompilationAvailable(project, projectChecksum));
+            }
+        }
 
-            lock (_lock)
+        private void OnCompilationAvailable(Project project, Checksum projectChecksum)
+        {
+            lock (_gate)
             {
-                if (isFullyLoaded && !_projectIdToCompilation.ContainsKey(project.Id))
-                {
-                    // If the project's compilation isn't yet available, kick off a task in the background to
-                    // hopefully make it available faster since we'll need it later to compute tokens.
-                    var newCompilationTask = project.GetCompilationAsync(cancellationToken);
-                    _projectIdToCompilation.Add(project.Id, newCompilationTask);
-                }
+                // Paranoia: It's technically possible (though unlikely) for two calls to compute the compilation for
+                // the same project to come back and call into this.  This is because the
+                // CompilationAvailableEventSource uses cooperative cancellation to cancel the in-flight request before
+                // issuing the new one.  There is no requirement though that the inflight request actually stop (or run
+                // to completion) prior to the next request running and completing.  In practice this should not happen
+                // as cancellation is checked fairly regularly.  However, if it does, check and do not bother to issue a
+                // refresh in this case.
+                if (ChecksumIsUnchanged_NoLock(project, projectChecksum))
```

Comment:
Maybe just making this a local function and calling it something like `IsChecksumUnchanged` would be the most concise? I feel like the naming of `ChecksumIsUnchanged_NoLock` is confusing without context. I'm guessing it just means that the method's implementation isn't protected by a lock, but without context and just looking at the name, I feel like it could also be interpreted as that the method shouldn't be called under a lock